File: Intersection of Two Linked Lists.java
// leetcode 160 problem
// 找出两个链表的公共部分（公共部分指中间某个结点至最后一个元素）

// 两个链表的长度可能不同，但是它们公共部分的长度相同，前面的长度不同，
// 我们用第一个循环消除长度差异，第二个循环找出公共部分
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        ListNode a = headA;
        ListNode b = headB;
        while (a != b) {
            a = a == null ? headB : a.next;
            b = b == null ? headA : b.next;
        }
        return a;
    }
}

File: Intersection of Two Linked Lists .cpp
// leetcode 160 problem
// 找出两个链表的公共部分

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (headA == NULL || headB == NULL) return NULL;
        int len1 = 0, len2 = 0;
        ListNode* l1 = headA;
        ListNode* l2 = headB;
        while (l1 && ++len1) { l1 = l1->next; }
        while (l2 && ++len2) { l2 = l2->next; }
        
        if (len1 > len2) return getIntersectionNode(headB, headA);
        
        int dif = len2 - len1;
        while (dif--) { headB = headB->next; }
        while (headA && headB) {
            if (headA == headB) return headA;
            headA = headA->next;
            headB = headB->next;
        }
        return NULL;
    }
};